lib/Utils/Layout/Utils.cpp

```diff
@@ -215,14 +240,93 @@ presburger::IntegerRelation getPerRowLayoutRelation(RankedTensorType matrixType,
   auto slotMod = addModConstraint(result, slotCoeffs, paddedCols);
 
   // slotMod - col = 0
-  SmallVector<int64_t> eqConstraint(result.getNumCols(), 0);
-  eqConstraint[slotMod] = 1;
-  eqConstraint[result.getVarKindOffset(VarKind::Domain) + 1] = -1;
-  result.addEquality(eqConstraint);
+  addConstraint(
+      result,
+      {{slotMod, 1}, {result.getVarKindOffset(VarKind::Domain) + 1, -1}},
+      /*equality=*/true);
 
   return result;
 }
 
+presburger::IntegerRelation get2dConvFilterRelation(RankedTensorType filterType,
+                                                    RankedTensorType dataType,
+                                                    int64_t padding) {
+  auto domainSize = filterType.getRank();
+  assert(domainSize == 2 && "expected 2-D filter matrix");
+
+  IntegerRelation result(PresburgerSpace::getRelationSpace(
+      domainSize, /*numRange=*/2, /*numSymbol=*/0, /*numLocals=*/2));
+
+  // These are the indices that represent the position of the filter as it
+  // slides over the data.
+  auto dataRow = result.getVarKindOffset(VarKind::Local);
+  auto dataCol = result.getVarKindOffset(VarKind::Local) + 1;
+
+  // Filter row and column indices
+  auto filterRow = result.getVarKindOffset(VarKind::Domain);
+  auto filterCol = result.getVarKindOffset(VarKind::Domain) + 1;
+
+  // Matrix row and column indices
+  auto matRow = result.getVarKindOffset(VarKind::Range);
+  auto matCol = result.getVarKindOffset(VarKind::Range) + 1;
+
+  // Constant coefficient
+  auto constCoeff = result.getNumCols() - 1;
+
+  // Filter and datasize
+  auto filterRowSize = filterType.getDimSize(0);
+  auto filterColSize = filterType.getDimSize(1);
+  auto dataRowSize = dataType.getDimSize(0);
+  auto dataColSize = dataType.getDimSize(1);
+
+  // Add bounds for the filter matrix dimensions.
+  addBounds(result, filterRow, 0, filterRowSize - 1);
+  addBounds(result, filterCol, 0, filterColSize - 1);
+
+  // Add bounds for the locals (the filter sliding indices).
+  addBounds(result, dataRow, -padding,
+            -padding + (2 * padding + dataRowSize - filterRowSize));
```

Comment:
```suggestion
            dataRowSize + padding - filterRowSize);
```

I kind of see why you wrote it this way, to identify the offset from the start (`-padding`), but I feel this is simpler and easier to understamd.